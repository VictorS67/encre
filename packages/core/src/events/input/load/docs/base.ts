import { BaseEvent, type BaseEventParams } from '../../../base.js';
import { type Context } from './context.js';

/**
 * Interface for defining a loader that is responsible for loading readable sources and returning them in a specified output format.
 * This loader is typically used for processing input data and transforming it into a structured context suitable for further processing.
 *
 * @template Input The type of input that the loader accepts.
 * @template Output The type of output that the loader produces, typically an array of Context objects.
 */
export interface DocLoader<Input = unknown, Output = Context[]> {
  /**
   * Loads data from the provided source and returns it as structured output.
   *
   * @param source The source input to load.
   * @returns A promise that resolves to the output, typically an array of contexts.
   */
  load(source: Input): Promise<Output>;
}

/**
 * Extends BaseEventParams to include options specific to document loaders, such as whether to split the context into chunks.
 * This interface is designed to provide customizable options for document loading processes.
 *
 * @extends BaseEventParams
 */
export interface BaseDocLoaderCallOptions extends BaseEventParams {
  /**
   * Optional flag indicating whether the loader should split the loaded context into separate chunks.
   * This can be useful for handling large datasets or distributing processing load.
   */
  shouldSplit?: boolean;
}

/**
 * Defines parameters for initializing BaseLoader instances, including all options from BaseDocLoaderCallOptions.
 * This interface is typically used to pass configuration options when creating new loader instances.
 *
 * @extends BaseDocLoaderCallOptions
 */
export interface BaseLoaderParams extends BaseDocLoaderCallOptions {}

/**
 * An abstract class that defines the framework for a document loader, capable of loading inputs and transforming them into a structured context.
 * Subclasses are expected to implement the specific logic for loading and processing the data based on the defined abstract methods.
 *
 * @template CallInput The type of input the loader accepts, defaulting to unknown.
 * @template CallOutput The type of output the loader produces, typically an array of Context objects.
 * @template CallOptions The options applicable to the loader, extending BaseDocLoaderCallOptions.
 */
export abstract class BaseLoader<
    CallInput = unknown,
    CallOutput = Context[],
    CallOptions extends BaseDocLoaderCallOptions = BaseDocLoaderCallOptions,
  >
  extends BaseEvent<CallInput, CallOutput, CallOptions>
  implements DocLoader<CallInput, CallOutput>
{
  /** @hidden */
  declare CallOptions: CallOptions;

  _eventNamespace(): string[] {
    return ['input', 'load', 'docs', this._docType()];
  }

  /**
   * Whether the loader should split the output into chunks.
   */
  shouldSplit?: boolean;

  constructor(fields?: BaseLoaderParams) {
    super(fields ?? {});

    this.shouldSplit = fields?.shouldSplit ?? false;
  }

  /**
   * Invokes the loading process with the given input and options.
   *
   * @param input The input data to load.
   * @param options Optional settings to customize the loading process.
   * @returns A Promise resolving to the output, typically an array of contexts.
   */
  async invoke(
    input: CallInput,
    options?: Partial<CallOptions>
  ): Promise<CallOutput> {
    this.shouldSplit = options?.shouldSplit ?? this.shouldSplit;

    return this.load(input);
  }

  /**
   * Defines the document type that this loader handles.
   *
   * @returns A string representing the document type.
   */
  abstract _docType(): string;

  /**
   * Define how sources are loaded.
   *
   * @param source The input source to be loaded.
   * @returns A Promise that resolves with the loaded output, as defined by CallOutput.
   */
  abstract load(source: CallInput): Promise<CallOutput>;
}
