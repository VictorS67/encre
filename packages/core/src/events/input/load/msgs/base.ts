import {
  Serializable,
  type SerializedConstructor,
} from '../../../../load/serializable.js';
import { isDeepEqual } from '../../../../utils/equal.js';
import {
  type SerializedMessage,
  type SerializedMessageData,
} from '../../../output/provide/serde.js';

/**
 * Represents the role of a message within a system, which helps in identifying the purpose and the origin of the message.
 * This role can influence how the message is processed or displayed in various parts of the system.
 */
export type MessageRole =
  | 'human' // Represents messages originating from human users.
  | 'assistant' // Represents messages generated by an automated assistant (e.g. LLMs).
  | 'system' // Represents prompt messages for LLMs.
  | 'function' // Represents messages that are the configuration of specific functions for function-calling/tool-calling in LLMs.
  | 'general'; // Represents general messages that do not fit into the predefined categories.

/**
 * Defines the structure of a typical message within the system, incorporating content and optional metadata such as role and name.
 */
export type Message = {
  /**
   * The main textual content of the message.
   */
  content: string;

  /**
   * Optional role of the message, which helps categorize the message's origin or purpose within the system.
   */
  role?: MessageRole;

  /**
   * An optional name associated with the message, which could be used for more specific identification or referencing purposes.
   */
  name?: string;
};

/**
 * Represents a flexible representation of a message, which can be either a complete BaseMessage instance, a tuple of role and string content,
 * or a simple string. This type is useful for functions that need to accept messages in various formats.
 */
export type BaseMessageLike = BaseMessage | [MessageRole, string] | string;

/**
 * Represents the types of content that can be used in messages. This type is designed to be flexible, 
 * allowing either a simple string or an object with key-value pairs, which can represent more structured data.
 */
export type ContentLike = string | { [key: string]: unknown };

/**
 * Defines the structure for the basic fields of a message within the system.
 * This interface includes the primary content of the message, an optional name,
 * and additional keyword arguments that might be used for further customization or processing.
 */
export interface BaseMessageFields {
  /**
   * The main content of the message. It can be a string or a more complex structure
   * represented by an object with key-value pairs.
   */
  content: ContentLike | ContentLike[];

  /**
   * An optional name for the message, which could be used for identifying or categorizing the message.
   */
  name?: string;

  /**
   * Optional additional keyword arguments that provide extended functionality and flexibility
   * in handling the message. These might include additional data fields not covered by the standard properties.
   */
  additionalKwargs?: {
    [key: string]: unknown;
  };
}

/**
 * An abstract base class for messages, implementing the BaseMessageFields interface and extending Serializable for data persistence.
 * BaseMessage provides foundational attributes and methods for message handling, serving as a blueprint for more specific message types.
 */
export abstract class BaseMessage
  extends Serializable
  implements BaseMessageFields
{
  _namespace: string[] = ['events', 'input', 'load', 'msgs', this._role()];

  /**
   * The main content of the message. It can be a string or a more complex structure
   * represented by an object with key-value pairs.
   */
  content: ContentLike | ContentLike[];

  /**
   * An optional name for the message, which could be used for identifying or categorizing the message.
   */
  name?: string;

  /**
   * Optional additional keyword arguments that provide extended functionality and flexibility
   * in handling the message. These might include additional data fields not covered by the standard properties.
   */
  additionalKwargs?: NonNullable<BaseMessageFields['additionalKwargs']>;

  /**
   * Abstract method to define the role of the message, which must be implemented by derived classes.
   * @returns The role of the message as a string.
   */
  abstract _role(): MessageRole;

  /**
   * Constructs a new instance of BaseMessage, initializing with the provided fields.
   * @param fields Either a string or an object conforming to BaseMessageFields to initialize the message content.
   */
  constructor(fields: string | BaseMessageFields) {
    if (typeof fields === 'string') {
      fields = { content: fields };
    }
    super(fields);

    this.name = fields.name;
    this.content = fields.content;
    this.additionalKwargs = fields.additionalKwargs;
  }

  /**
   * Serializes the message to a structured format that includes the role and the JSON-converted content.
   * @returns An object representing the serialized state of the message.
   */
  toSerialized(): SerializedMessage {
    return {
      role: this._role(),
      json: (this.toJSON() as SerializedConstructor)
        ._kwargs as object as SerializedMessageData,
    };
  }

  /**
   * Coerce the message content to boolean.
   *
   * Usecases if the message content is coerceable:
   * - if content equals to either "true" or "false", return corresponding boolean;
   * - `true` if content length larger than 0, otherwise return `false`;
   * - `true` if content is not string type;
   * - if content is array, return `true` if all entries follows the previous three usecases, otherwise return `false`.
   *
   * @returns boolean if coerceable, otherwise return undefined.
   */
  coerceToBoolean(): boolean | undefined {
    const coerceContentLikeToBoolean = (contentLike: ContentLike): boolean => {
      if (typeof contentLike === 'string') {
        return contentLike.length > 0 && contentLike.toLowerCase() !== 'false';
      }

      return true;
    };

    if (Array.isArray(this.content)) {
      return (
        this.content.length > 0 &&
        this.content.every((c) => coerceContentLikeToBoolean(c))
      );
    }

    return coerceContentLikeToBoolean(this.content);
  }

  /**
   * Abstract method to concatenate this message with another message, producing a new BaseMessage instance.
   *
   * @param message The message to concatenate with this instance.
   * @returns A new instance of BaseMessage resulting from the concatenation.
   */
  abstract concat(message: BaseMessage): BaseMessage;

  /**
   * Internal method for merging two message conents.
   *
   * @param left left message content.
   * @param right right message content.
   * @returns merged message content.
   * @internal
   */
  static _mergeContent(
    left: NonNullable<BaseMessageFields['content']>,
    right: NonNullable<BaseMessageFields['content']>
  ): NonNullable<BaseMessageFields['content']> {
    if (!Array.isArray(left) && !Array.isArray(right)) {
      if (typeof left === 'string' && typeof right === 'string') {
        return left + right;
      }

      if (typeof left === 'object' && typeof right === 'object') {
        return {
          ...left,
          ...right,
        };
      }

      throw new Error(
        'Message cannot be merged. Content should follow the same type.'
      );
    }

    if (!Array.isArray(left) && Array.isArray(right)) {
      return [left, ...right];
    }

    if (Array.isArray(left) && !Array.isArray(right)) {
      return [...left, right];
    }

    if (Array.isArray(left) && Array.isArray(right)) {
      return [...left, ...right];
    }

    throw new Error(
      'Message cannot be merged. Content should follow the same type.'
    );
  }

  /**
   * Internal method for merging two additional keyword arguments.
   *
   * @param left left additional keyword arguments.
   * @param right right additional keyword arguments.
   * @returns merged additional keyword arguments.
   * @internal
   */
  static _mergeAdditionalKwargs(
    left: NonNullable<BaseMessageFields['additionalKwargs']>,
    right: NonNullable<BaseMessageFields['additionalKwargs']>
  ): NonNullable<BaseMessageFields['additionalKwargs']> {
    const additionalKwargs = { ...left };
    const newKwargs = { ...right };

    for (const [k, v] of Object.entries(newKwargs)) {
      if (typeof additionalKwargs[k] !== typeof v) {
        throw new Error(
          `key ${k} in additionalKwargs but values do not match types.`
        );
      } else if (additionalKwargs[k] === undefined) {
        additionalKwargs[k] = v;
      } else if (typeof additionalKwargs[k] === 'string') {
        additionalKwargs[k] = (additionalKwargs[k] as string) + v;
      } else if (
        !Array.isArray(additionalKwargs[k]) &&
        typeof additionalKwargs[k] === 'object'
      ) {
        additionalKwargs[k] = this._mergeAdditionalKwargs(
          additionalKwargs[k] as NonNullable<
            BaseMessageFields['additionalKwargs']
          >,
          v as NonNullable<BaseMessageFields['additionalKwargs']>
        );
      } else {
        throw new Error(
          `key ${k} already in additionalKwargs but failed to merge.`
        );
      }
    }

    return additionalKwargs;
  }

  /**
   * Check if two messages are strictly equal.
   *
   * @param message1 message object.
   * @param message2 message object.
   * @returns true if strictly equal.
   */
  static isEqualMessage(message1: BaseMessage, message2: BaseMessage) {
    return isDeepEqual(message1.toSerialized(), message2.toSerialized());
  }

  /**
   * Check if two message arrays are strictly equal.
   *
   * @param messageArray1 message object array.
   * @param messageArray2 message object array.
   * @returns true if strictly equal.
   */
  static isEqualMessageArray(
    messageArray1: BaseMessage[],
    messageArray2: BaseMessage[]
  ) {
    if (messageArray1.length !== messageArray2.length) {
      return false;
    }

    for (let i = 0; i < messageArray1.length; i++) {
      if (!this.isEqualMessage(messageArray1[i], messageArray2[i])) {
        return false;
      }
    }

    return true;
  }
}
