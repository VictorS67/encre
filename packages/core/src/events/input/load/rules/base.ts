import { type RecordId } from '../../../../load/keymap.js';
import { Serializable } from '../../../../load/serializable.js';
import { type SerializedRuleCollection } from '../../../../serde.js';
import { BaseRule } from '../../../inference/validate/guardrails/index.js';

/**
 * Defines the structure for initializing a BaseRuleCollection.
 * It specifies a collection of rules and optionally a conjunction to define how
 * the rules should be evaluated.
 */
export interface BaseRuleCollectionField {
  /**
   * A collection of rules, each identified by a string key. The value can be
   * either a BaseRule or another BaseRuleCollection.
   */
  collection: Record<string, BaseRule | BaseRuleCollection>;

  /**
   * The logical conjunction used to evaluate the rules. Default is "and".
   * Possible values are 'and' or 'or'.
   */
  conjunction?: 'and' | 'or';
}

/**
 * A collection of rules that can be evaluated together. This provides functionality
 * to validate a set of inputs against the rules defined in the collection.
 */
export class BaseRuleCollection
  extends Serializable
  implements BaseRuleCollectionField
{
  _isSerializable = true;

  _namespace: string[] = ['events', 'input', 'load', 'rules'];

  /**
   * A collection of rules, each identified by a string key. The value can be
   * either a BaseRule or another BaseRuleCollection.
   */
  collection: Record<string, BaseRule | BaseRuleCollection>;

  /**
   * The logical conjunction used to evaluate the rules. Default is "and".
   * Possible values are 'and' or 'or'.
   */
  conjunction: 'and' | 'or';

  /**
   * Initializes a new instance of BaseRuleCollection with the specified rules
   * and conjunction.
   * @param fields The fields required to initialize the rule collection.
   */
  constructor(fields: BaseRuleCollectionField) {
    super(fields);

    this.collection = fields.collection;
    this.conjunction = fields.conjunction ?? 'and';
  }

  /**
   * Provides a human-readable description of the rule collection, detailing the
   * rules and their logical relationship.
   */
  get description(): string {
    return Object.entries(this.collection)
      .map(([subject, rule]) => {
        const currDescription: string = rule.description;
        const isInvert: boolean = this._isInvertLogic(
          currDescription,
          this.conjunction
        );
        const isCollection: boolean = rule instanceof BaseRuleCollection;

        return isInvert
          ? `(${
              !isCollection ? `{{{${subject}}}} ` : ''
            }<${subject}>${currDescription}</${subject}>)`
          : `${
              !isCollection ? `{{{${subject}}}} ` : ''
            }<${subject}>${currDescription}</${subject}>`;
      })
      .join(` ${this.conjunction.toUpperCase()} `);
  }

  /**
   * Deserializes a given serialized rule collection into an instance of BaseRuleCollection.
   * @param serialized The serialized rule collection.
   * @param values Optional additional values for rule deserialization.
   * @returns The deserialized rule collection.
   */
  static async deserialize(
    serialized: SerializedRuleCollection,
    values: Record<string, unknown> = {}
  ): Promise<BaseRuleCollection> {
    if (serialized._type !== 'rule-collection') {
      throw new Error(
        `CANNOT deserialize this type in rule collection: ${serialized._type}`
      );
    }

    const { BaseRule } = await import(
      '../../../inference/validate/guardrails/base.js'
    );

    const ruleCollectionField: BaseRuleCollectionField = {
      collection: Object.fromEntries(
        await Promise.all(
          Object.entries(serialized.collection).map(async ([subject, rule]) => [
            subject,
            rule._type === 'rule-collection'
              ? await BaseRuleCollection.deserialize(
                  rule,
                  (values?.[subject] ?? {}) as Record<string, unknown>
                )
              : await BaseRule.deserialize(
                  rule,
                  (values?.[subject] ?? {}) as Record<string, unknown>
                ),
          ])
        )
      ),
      conjunction: serialized.conjunction,
    };

    return new BaseRuleCollection(ruleCollectionField);
  }

  /**
   * Serializes the current state of the rule collection to a format suitable for storage or transmission.
   * This method facilitates the conversion of the rule collection into a simplified JSON format,
   * capturing essential properties like description, collection, and conjunction.
   * @returns The serialized form of the rule collection.
   */
  serialize(): SerializedRuleCollection {
    return {
      _type: 'rule-collection',
      description: this.description,
      collection: Object.fromEntries(
        Object.entries(this.collection).map(([subject, rule]) => [
          subject,
          rule.serialize(),
        ])
      ),
      conjunction: this.conjunction,
    };
  }

  /**
   * Generates a simplified description of the rule collection without any special formatting.
   * This method provides a clean, readable description of each rule's logic without including
   * XML-like tags or additional decorators used in more complex descriptions.
   * @returns A simplified string representation of the rule collection's logic.
   */
  getCleanDescription(): string {
    return Object.entries(this.collection)
      .map(([subject, rule]) => {
        const currDescription: string = rule.getCleanDescription();
        const isInvert: boolean = this._isInvertLogic(
          currDescription,
          this.conjunction
        );
        const isCollection: boolean = rule instanceof BaseRuleCollection;

        return isInvert
          ? `(${!isCollection ? `${subject} ` : ''}${currDescription})`
          : `${!isCollection ? `${subject} ` : ''}${currDescription}`;
      })
      .join(` ${this.conjunction.toUpperCase()} `);
  }

  /**
   * Validates the provided inputs against the rules in the collection.
   * @param input Inputs to validate.
   * @param variables Optional variables for rule validation.
   * @returns The result of the validation.
   */
  async validate(
    input: {
      [key in string]: unknown;
    },
    variables?: {
      [key in string]?: Record<string, unknown>;
    }
  ): Promise<boolean> {
    if (
      !this._isIdenticalKeySet(Object.keys(this.collection), Object.keys(input))
    ) {
      throw new Error(
        'CANNOT perform validation in rule collection because of two different key sets.'
      );
    }

    const recordHistory: Set<RecordId> = new Set([this._recordId]);

    if (Object.keys(this.collection).length === 0) {
      return false;
    }

    if (this.conjunction === 'and') {
      return Object.entries(this.collection).every(async ([subject, rule]) => {
        return this._validate(subject, rule, recordHistory, input, variables);
      });
    }

    return Object.entries(this.collection).some(async ([subject, rule]) => {
      return this._validate(subject, rule, recordHistory, input, variables);
    });
  }

  /**
   * Combines two rule collections into a new rule collection with a specified conjunction.
   * @param ruleCollection Another rule collection to combine with this one.
   * @param conjunction The conjunction to use for the new combined rule collection.
   * @returns The new combined rule collection.
   */
  concat(
    ruleCollection: BaseRuleCollection,
    conjunction: 'and' | 'or'
  ): BaseRuleCollection {
    return new BaseRuleCollection({
      collection: {
        '0': this,
        '1': ruleCollection,
      },
      conjunction,
    });
  }

  /**
   * Validates a specific input against a specific rule within the rule collection.
   * This protected method is used internally by the `validate` method to apply individual
   * rules  to corresponding parts of the input data.
   * @param subject The key representing the specific part of input data to validate.
   * @param rule The rule or rule collection to apply.
   * @param recordHistory A set of record IDs to track rule application and prevent recursion.
   * @param input The entire input data.
   * @param variables Optional additional variables needed for validation.
   * @returns The result of the validation for the specific input part.
   * @internal
   */
  protected async _validate(
    subject: string,
    rule: BaseRule | BaseRuleCollection,
    recordHistory: Set<RecordId>,
    input: {
      [key in string]: unknown;
    },
    variables?: {
      [key in string]?: Record<string, unknown>;
    }
  ): Promise<boolean> {
    if (rule instanceof BaseRule) {
      return await rule.validate(input[subject], variables?.[subject]);
    }
    const collectionVariables = Object.fromEntries(
      Object.entries(variables ?? {})
        .filter(([cSubject, _]) => {
          return cSubject.startsWith(`${subject}.`);
        })
        .map(([cSubject, cVariables]) => {
          const substring = `${subject}.`;
          const startIndex = cSubject.indexOf(substring);

          if (startIndex === -1) {
            throw new Error(
              `CANNOT perform validation bacause of incorrect key name of the variable: ${subject}`
            );
          }

          const sliceFromIndex = startIndex + substring.length;
          return [cSubject.substring(sliceFromIndex), cVariables];
        })
    );

    if (recordHistory.size === 17) {
      throw new Error('number of collections reaches the maximum');
    }

    if (recordHistory.has(rule._recordId)) {
      throw new Error('CANNOT recursively register rule collection');
    }
    recordHistory.add(rule._recordId);

    return await rule.validate(input, collectionVariables);
  }

  protected _isIdenticalKeySet(keyset1: string[], keyset2: string[]): boolean {
    if (keyset1.length !== keyset2.length) {
      return false;
    }

    const sortedKeys1 = keyset1.slice().sort();
    const sortedKeys2 = keyset1.slice().sort();

    for (let i = 0; i < sortedKeys1.length; i++) {
      if (sortedKeys1[i] !== sortedKeys2[i]) {
        return false;
      }
    }

    return true;
  }

  /**
   * Determines whether a given description represents a composite rule.
   * This method checks if the description includes logical operators such as 'AND' or 'OR',
   * indicating a composite of multiple rules.
   * @param description The rule description to check.
   * @returns True if the description includes composite rule indicators, false otherwise.
   * @internal
   */
  protected _isCompositeRule(description: string): boolean {
    return description.includes('AND') || description.includes('OR');
  }

  /**
   * Extracts the outermost logical conjunction ('AND' or 'OR') from a rule description.
   * This method analyzes the rule description to determine the highest-level logical operator
   * governing its logic.
   * @param description The rule description to analyze.
   * @returns The outermost logical conjunction ('AND' or 'OR').
   * @throws Throws if the description does not contain a clear outermost conjunction.
   * @internal
   */
  protected _getOutermostConjunction(description: string): string {
    const isMatchingParentheses = (expression: string): boolean => {
      let balance = 0;
      for (let i = 0; i < expression.length; i++) {
        if (expression[i] === '(') balance++;
        else if (expression[i] === ')') balance--;

        if (balance < 0) return false;
      }
      return balance === 0;
    };

    let expression = description.trim();

    while (
      expression.startsWith('(') &&
      expression.endsWith(')') &&
      isMatchingParentheses(expression)
    ) {
      expression = expression.substring(1, expression.length - 1).trim();
    }

    let depth = 0;
    let cleanedExpression = '';

    for (let i = 0; i < expression.length; i++) {
      const char = expression[i];
      if (char === '(') {
        depth++;
      } else if (char === ')') {
        depth--;
      } else if (depth === 0) {
        cleanedExpression += char;
      }
    }

    const andIndex: number = cleanedExpression.indexOf('AND');
    const orIndex: number = cleanedExpression.indexOf('OR');

    if (andIndex !== -1 && (andIndex < orIndex || orIndex === -1)) {
      return 'AND';
    } else if (orIndex !== -1 && (orIndex < andIndex || andIndex === -1)) {
      return 'OR';
    }

    throw new Error(
      `description CANNOT get outmost conjunction: ${description}`
    );
  }

  /**
   * Checks if the logic of a rule description is inverted relative to a given conjunction.
   * This method is used to determine if the description's logic needs to be presented 
   * differently based on the specified conjunction.
   * @param description The description of the rule.
   * @param conjunction The conjunction to compare against.
   * @returns True if the rule's logic is inverted relative to the conjunction, false otherwise.
   * @internal
   */
  protected _isInvertLogic(
    description: string,
    conjunction: 'and' | 'or'
  ): boolean {
    if (!this._isCompositeRule(description)) {
      return false;
    }

    const currConjunction: string = this._getOutermostConjunction(description);
    return currConjunction.toLowerCase() !== conjunction;
  }
}
