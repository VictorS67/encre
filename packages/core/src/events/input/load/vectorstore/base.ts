import { Serializable } from '../../../../load/serializable.js';
import { type Context } from '../docs/index.js';

export interface BaseVectorStoreField {}

/**
 * An abstract class representing the basic functionality of a vector store. Vector stores
 * are used to manage and retrieve high-dimensional vector data typically used in search
 * and machine learning applications. This class provides a common interface for adding,
 * deleting, and searching vectors within a store.
 *
 * @extends Serializable Indicates that instances of subclasses can be serialized.
 * @implements {BaseVectorStoreField} A marker interface for additional fields.
 */
export abstract class BaseVectorStore
  extends Serializable
  implements BaseVectorStoreField
{
  /**
   * Defines the filter type used in vector operations. This can be an object or a string,
   * depending on the subclass implementation, to support various filtering strategies.
   */
  declare FilterType: object | string;

  _isSerializable = false;

  _namespace: string[] = [
    'events',
    'input',
    'load',
    'vectorstore',
    this._vectorstoreType(),
  ];

  constructor(fields?: Partial<BaseVectorStoreField>) {
    super(fields);
  }

  /**
   * Returns the type of the vector store.
   */
  abstract _vectorstoreType(): string;

  /**
   * Adds vectors to the store along with their associated context data.
   * @param embeddings An array of numerical arrays representing the vector embeddings to be added.
   * @param context An array of `Context` objects associated with each embedding.
   * @param options Optional parameters for the add operation.
   * @returns {Promise<void>} A promise that resolves when the vectors have been added.
   */
  abstract addVectors(
    embeddings: number[][],
    context: Context[],
    options?: Record<string, unknown>
  ): Promise<void>;

  /**
   * Deletes vectors from the store based on specified parameters.
   * @param params Optional parameters to specify which vectors to delete.
   * @returns {Promise<void>} A promise that resolves when the vectors have been deleted.
   */
  abstract deleteVectors(params?: Record<string, unknown>): Promise<void>;

  /**
   * Performs a similarity search to find vectors closest to the given query vector.
   * @param query A numerical array representing the query vector.
   * @param topK The number of top similar results to return.
   * @param filter Optional filter to apply to the search.
   * @returns {Promise<[Context, number][]>} A promise that resolves to an array of tuples,
   *         each containing a `Context` object and a similarity score.
   */
  abstract similaritySearch(
    query: number[],
    topK: number,
    filter?: this['FilterType']
  ): Promise<[Context, number][]>;

  /**
   * Performs a Max Marginal Relevance (MMR) search to diversify the results of a similarity search.
   * This method helps in retrieving not just the most similar items, but also the most diverse
   * ones from the top results.
   * @param query A numerical array representing the query vector.
   * @param topK The number of top results to consider for diversification.
   * @param lambda A parameter controlling the trade-off between similarity and diversity.
   * @param filter Optional filter to apply during the search.
   * @returns {Promise<[Context, number][]>} A promise that resolves to an array of tuples,
   *         each containing a `Context` object and a relevance score.
   */
  abstract maxMarginalRelevanceSearch(
    query: number[],
    topK: number,
    lambda: number,
    filter?: this['FilterType']
  ): Promise<[Context, number][]>;
}
