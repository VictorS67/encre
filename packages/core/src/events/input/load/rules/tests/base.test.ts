import { describe, expect, test } from '@jest/globals';
import { ArrayRule } from '../../../../inference/validate/guardrails/array.js';
import { StringRule } from '../../../../inference/validate/guardrails/string.js';
import { BaseRuleCollection } from '../base.js';

describe('test BaseRuleCollection', () => {
  test('simple RuleCollection', () => {
    const stringExistsRule = StringRule.exists();
    const stringEqualsToRule = StringRule.isEqual('John');
    const arrayContainsRule = ArrayRule.contains('Peter');
    const stringLengthLargerRule = StringRule.lengthGreaterThan(2);

    const ruleCollection1 = new BaseRuleCollection({
      collection: {
        A: stringExistsRule
          .concat(stringEqualsToRule, 'and')
          .concat(stringLengthLargerRule, 'or'),
        B: arrayContainsRule,
      },
    });

    const ruleCollection2 = new BaseRuleCollection({
      collection: {
        C: stringLengthLargerRule,
      },
    });

    const ruleCollection = ruleCollection1.concat(ruleCollection2, 'or');

    expect(ruleCollection.serialize()).toMatchSnapshot();
  });
});
