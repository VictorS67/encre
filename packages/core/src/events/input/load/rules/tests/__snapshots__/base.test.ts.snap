// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test BaseRuleCollection simple RuleCollection 1`] = `
{
  "_type": "rule-collection",
  "collection": {
    "0": {
      "_type": "rule-collection",
      "collection": {
        "A": {
          "_ruleType": "string",
          "_type": "rule",
          "description": "(exists AND is equal to {{right.value}}) OR length greater than {{right.value}}",
          "func": "function anonymous(input,variables
) {
return (async function(input, variables) {

  const leftFunc = async (input, variables) => {

return (async function(input, variables) {

  const leftFunc = async (input) => {

return (async function(input) {
                return input !== undefined;
            }).apply(this, [input])

};
  const rightFunc = async (input, variables) => {

return (async function(input, variables) {
                return input === variables.value;
            }).apply(this, [input, variables])

};

  const result1 = await leftFunc(input).catch(
    (e) => false
  );
  const result2 = await rightFunc(input, variables.right).catch(
    (e) => false
  );

  const results = await Promise.all([result1, result2]);

  return results.every((result) => result === true);

}).apply(this, [input, variables])

};
  const rightFunc = async (input, variables) => {

return (async function(input, variables) {
                return input.length > variables.value;
            }).apply(this, [input, variables])

};

  const result1 = await leftFunc(input, variables.left).catch(
    (e) => false
  );
  const result2 = await rightFunc(input, variables.right).catch(
    (e) => false
  );

  const results = await Promise.all([result1, result2]);

  return results.some((result) => result === true);

}).apply(this, [input, variables])
}",
          "metadata": {
            "conjunction": "or",
            "left": {
              "_ruleType": "string",
              "_type": "rule",
              "description": "exists AND is equal to {{right.value}}",
              "func": "function anonymous(input,variables
) {
return (async function(input, variables) {

  const leftFunc = async (input) => {

return (async function(input) {
                return input !== undefined;
            }).apply(this, [input])

};
  const rightFunc = async (input, variables) => {

return (async function(input, variables) {
                return input === variables.value;
            }).apply(this, [input, variables])

};

  const result1 = await leftFunc(input).catch(
    (e) => false
  );
  const result2 = await rightFunc(input, variables.right).catch(
    (e) => false
  );

  const results = await Promise.all([result1, result2]);

  return results.every((result) => result === true);

}).apply(this, [input, variables])
}",
              "metadata": {
                "conjunction": "and",
                "left": {
                  "_ruleType": "string",
                  "_type": "rule",
                  "description": "exists",
                  "func": "function anonymous(input
) {
return (async function(input) {
                return input !== undefined;
            }).apply(this, [input])
}",
                  "metadata": undefined,
                  "variables": undefined,
                },
                "right": {
                  "_ruleType": "string",
                  "_type": "rule",
                  "description": "is equal to {{value}}",
                  "func": "function anonymous(input,variables
) {
return (async function(input, variables) {
                return input === variables.value;
            }).apply(this, [input, variables])
}",
                  "metadata": undefined,
                  "variables": {
                    "value": "John",
                  },
                },
              },
              "variables": {
                "left": {},
                "right": {
                  "value": "John",
                },
              },
            },
            "right": {
              "_ruleType": "string",
              "_type": "rule",
              "description": "length greater than {{value}}",
              "func": "function anonymous(input,variables
) {
return (async function(input, variables) {
                return input.length > variables.value;
            }).apply(this, [input, variables])
}",
              "metadata": undefined,
              "variables": {
                "value": 2,
              },
            },
          },
          "variables": {
            "left": {
              "left": {},
              "right": {
                "value": "John",
              },
            },
            "right": {
              "value": 2,
            },
          },
        },
        "B": {
          "_ruleType": "array",
          "_type": "rule",
          "description": "contains {{value}}",
          "func": "function anonymous(input,variables
) {
return (async function(input, variables) {
                return input.includes(variables.value);
            }).apply(this, [input, variables])
}",
          "metadata": undefined,
          "variables": {
            "value": "Peter",
          },
        },
      },
      "conjunction": "and",
      "description": "({{{A}}} <A>(exists AND is equal to {{right.value}}) OR length greater than {{right.value}}</A>) AND {{{B}}} <B>contains {{value}}</B>",
    },
    "1": {
      "_type": "rule-collection",
      "collection": {
        "C": {
          "_ruleType": "string",
          "_type": "rule",
          "description": "length greater than {{value}}",
          "func": "function anonymous(input,variables
) {
return (async function(input, variables) {
                return input.length > variables.value;
            }).apply(this, [input, variables])
}",
          "metadata": undefined,
          "variables": {
            "value": 2,
          },
        },
      },
      "conjunction": "and",
      "description": "{{{C}}} <C>length greater than {{value}}</C>",
    },
  },
  "conjunction": "or",
  "description": "(<0>({{{A}}} <A>(exists AND is equal to {{right.value}}) OR length greater than {{right.value}}</A>) AND {{{B}}} <B>contains {{value}}</B></0>) OR <1>{{{C}}} <C>length greater than {{value}}</C></1>",
}
`;
